provider/core/src/marker.rs

```diff
@@ -108,6 +108,45 @@ impl<M: DataMarker + Sized> DataMarkerExt for M {
     }
 }
 
+/// This trait should be implemented on all data structs.
+///
+/// For most data structs, a default implementation available via
+/// [`does_not_deref_to_varule`] can be used.
+pub trait MaybeAsVarULE {
+    /// The [`VarULE`] type for this data struct.
+    type VarULE: VarULE + ?Sized;
+
+    /// If `self` can be wholy represented as [`Self::VarULE`],
+    /// returns it as a reference.
+    /// 
+    /// Otherwise, returns `None`.
+    fn maybe_as_varule(&self) -> Option<&Self::VarULE>;
+}
+
+/// Implements [`MaybeAsVarULE`] on a type that is NOT representable
```

Comment:
thought: the reverse default might be better UX for ICU4X devs so they do not accidentally feel forced to structure all of their data structs to support VarULE. I know Younies has found this confusing before.

but I figure there's a reason behind this choice here. if it's just macroability I think the macro can be done differently.

Reply:
> the reverse default

I don't understand what you mean. Do you mean the name of the trait? The name of the macro? Something else?

Reply:
This macro opts out of this system by default.

I'd rather have the data marker macro opt in by default, with a flag to opt out

Reply:
Not the name, just the default behavior, yes.

Reply:
Basically: if I define a new data struct, the default behavior of the data struct/marker macros should not make me think I need to impl VarULE. The non-None impl of this trait should be opt in.

Reply:
The macro does what I think you want it to do? It implements the trait to return None. If you want not None, there is no macro; implement the trait manually.

Reply:
I applied @robertbastian's suggestion to create a new `icu_provider::marker::data_struct!` macro rules.